certbot/interfaces.py

```diff
@@ -604,3 +604,204 @@ def renew_deploy(self, lineage, *args, **kwargs):
+
+    @abc.abstractmethod
+    def dirty(self):
+        """
+        Returns a boolean telling if the state of the node was changed after the
+        initial read or the last save operation. The main usage for this method is
+        to discover the nodes in the tree that need to be written to disk upon
+        save.
+
+        :param str name: The name of the child node to add
+
+        :returns: True if the node needs to be written to disk
+        """
+
+    @abc.abstractmethod
+    def save(self, msg):
+        """
+        Save traverses the children, and attempts to write the AST to disk for
+        all the objects that are marked dirty. The actual operation of course
+        depends on the underlying implementation. save() shouldn't be called
+        from the Configurator outside of it's designated save() method in order
+        to ensure that the Reverter checkpoints are created properly.
+
+        Note: this approach of keeping internal structure of the configuration
+        within the ParserNode tree does not represent the file inclusion structure
+        of actual configuration files that reside in the filesystem. To handle
+        file writes properly, the file specific temporary trees should be extracted
+        from the full ParserNode tree where necessary when writing to disk.
+
+        """
+
+    @abc.abstractmethod
+    def set_arguments(self, arguments):
+        """
+        Sets argument list of the ParserNode object and marks the node dirty.
+        """
+
+
+@six.add_metaclass(abc.ABCMeta)
+class CommentNode(ParserNode):
+    """
+    CommentNode class is used for representation of comments within the parsed
+    configuration structure. Because of the nature of comments, it is not able
+    to have child nodes and hance it is always treated as a leaf node.
+    """
+
+
+@six.add_metaclass(abc.ABCMeta)
+class DirectiveNode(ParserNode):
+    """
+    DirectiveNode class represents a configuration directive within the configuration.
+    It can have zero or more arguments attached to it. Because of the nature of
+    single directives, it is not able to have child nodes and hence it is always
+    treated as a leaf node.
+    """
+
+
+@six.add_metaclass(abc.ABCMeta)
+class BlockNode(ParserNode):
+    """
+    BlockNode class represents a block of nested configuration directives, comments
+    and other blocks as its children. A BlockNode can have zero or more arguments
+    attached to it.
+
+    Configuration blocks typically consist of one or more child nodes of all possible
+    types. Because of this, the BlockNode class has various discovery and structure
+    management methods.
+    """
+
+    @abc.abstractmethod
+    def add_child_block(self, name, arguments=[], position=-1):
+        """
+        Adds a new BlockNode child node with provided values and marks the callee
+        BlockNode dirty. This is used to add new children to the AST.
+
+        :param str name: The name of the child node to add
```

Comment:
What is the meaning of this `name` parameter for each of block, directive, and comments? Does the `name` also appear in the list of arguments? If so, is it ever not the first argument?

Reply:
Name is the directive "key", for `<VirtualHost *:80>` the name would be `VirtualHost`, for `Listen 80` the name would be `Listen`. It's not in the argument list.

Reply:
Do comments have a meaningful `name`? If not, I think we should remove the parameter from `add_child_comment` and only have a `name` attribute for block and directive nodes.

Reply:
SGTM